### ISSUE TITLE
0 byte file uploaded

### ISSUE THREAD

**Issue by chaoyangnz** [egnyte/egnyte-js-sdk#11]
I was using your nodejs SDK to upload with `storeFile` or `chunkStream`, both can cause 0 byte file uploaded.

I saw a warning for this case, `developer over QPS, retrying`. other than that, there is no error thrown, but randomly some 0 byte files are uploaded.

**Comment by uhlryk** [comment]
@chaoyangnz, please provide more details about this issue? (your domain, a few timestamps when the issue happened, file size, extensions) 
How often this issue happens?
How does the process look like? Do you stream files from other services or from the local file system?
Could you provide a working code snippet that fails with this 0-byte issue? 

`developer over QPS, retrying` should not be a case. Egnyte API has QPS, and SDK handles this limitation.

**Comment by chaoyangnz** [comment]
@uhlryk I think there is a defect in SDK to handle retry, after each retry, the request body size reduced.


the below is our log:

```
info: 🔵 POST https://rakentesting.egnyte.com/pubapi/v1/fs-content/Shared/Raken/chao_egnyte_test_project/Raken%20Gallery/01%20BiznessName1/2020/10.%20October/sample.2020.10.15_19.30.40.mp4 
{
    "0": "ba75d6d9-7666-4bd2-80c6-df97f2ea80a8",
    "1": {
        "url": "https://rakentesting.egnyte.com/pubapi/v1/fs-content/Shared/Raken/chao_egnyte_test_project/Raken%20Gallery/01%20BiznessName1/2020/10.%20October/sample.2020.10.15_19.30.40.mp4",
        "method": "POST",
        "headers": {
            "authorization": "Bearer 8sd******************xwq",
            "host": "rakentesting.egnyte.com",
            "content-type": "video/mp4",
            "x-honeycomb-trace": "1;trace_id=23786d4b-14e7-456f-a320-8e5c723b4978,parent_id=8cec6f6903c5a292,dataset=dev,context=e30=",
            "user-agent": "raken-integrations"
        },
        "body": "<383631 bytes binary>"
    },
    "service": "raken-integrations",
    "host": "169.254.22.9",
    "timestamp": "2021-04-14T01:59:05.221Z"
}
info: 🟡 403 POST https://rakentesting.egnyte.com/pubapi/v1/fs-content/Shared/Raken/chao_egnyte_test_project/Raken%20Gallery/01%20BiznessName1/2020/10.%20October/sample.2020.10.15_19.30.40.mp4 (119 ms) {
    "0": "ba75d6d9-7666-4bd2-80c6-df97f2ea80a8",
    "1": {
        "url": "https://rakentesting.egnyte.com/pubapi/v1/fs-content/Shared/Raken/chao_egnyte_test_project/Raken%20Gallery/01%20BiznessName1/2020/10.%20October/sample.2020.10.15_19.30.40.mp4",
        "method": "POST",
        "headers": {
            "authorization": "Bearer 8sd******************xwq",
            "host": "rakentesting.egnyte.com",
            "content-type": "video/mp4",
            "x-honeycomb-trace": "1;trace_id=23786d4b-14e7-456f-a320-8e5c723b4978,parent_id=8cec6f6903c5a292,dataset=dev,context=e30=",
            "user-agent": "raken-integrations"
        },
        "body": "<383631 bytes binary>"
    },
    "2": {
        "statusCode": 403,
        "statusMessage": "Forbidden",
        "headers": {
            "x-mashery-responder": "sjc-gcp-mashery443-slave01",
            "x-accesstoken-qps-allotted": "2",
            "x-accesstoken-qps-current": "5",
            "retry-after": "1",
            "x-error-detail-header": "Account Over Queries Per Second Limit",
            "x-mashery-error-code": "ERR_403_DEVELOPER_OVER_QPS",
            "content-type": "text/xml",
            "date": "Wed, 14 Apr 2021 01:59:12 GMT",
            "content-length": "27",
            "connection": "close",
            "x-content-type-options": "nosniff",
            "x-xss-protection": "1; mode=block",
            "strict-transport-security": "max-age=31536000; includeSubDomains",
            "x-egnyte-request-id": "3426FE37:4341_23F4C2EA:01BB_--_1EADE7|sjc-gcp-glb-l1webui03+http_l1_webui",
            "egnyte-processing-location": "GCP"
        },
        "body": "<27 bytes binary>"
    },
    "service": "raken-integrations",
    "host": "169.254.22.9",
    "timestamp": "2021-04-14T01:59:05.221Z"
}
2021-04-14T01:59:12.560Z	4d82ade8-e019-5427-8ac4-f518aea99fc7	WARN	developer over QPS, retrying
info: 🔵 POST https://rakentesting.egnyte.com:443/pubapi/v1/fs-content/Shared/Raken/chao_egnyte_test_project/Raken%20Gallery/01%20BiznessName1/2020/10.%20October/sample.2020.10.15_19.30.40.mp4 {
    "0": "b27f2ed8-5d22-495d-8dfc-823b64963af9",
    "1": {
        "url": "https://rakentesting.egnyte.com:443/pubapi/v1/fs-content/Shared/Raken/chao_egnyte_test_project/Raken%20Gallery/01%20BiznessName1/2020/10.%20October/sample.2020.10.15_19.30.40.mp4",
        "method": "POST",
        "headers": {
            "authorization": "Bearer 8sd******************xwq",
            "host": "rakentesting.egnyte.com",
            "content-length": 0,
            "user-agent": "raken-integrations"
        },
        "body": "<0 bytes binary>"
    },
    "service": "raken-integrations",
    "host": "169.254.22.9",
    "timestamp": "2021-04-14T01:59:05.221Z"
}
info: 🟡 403 POST https://rakentesting.egnyte.com:443/pubapi/v1/fs-content/Shared/Raken/chao_egnyte_test_project/Raken%20Gallery/01%20BiznessName1/2020/10.%20October/sample.2020.10.15_19.30.40.mp4 (75 ms) {
    "0": "59e95164-030a-470a-a2c8-9efb2732aba6",
    "1": {
        "url": "https://rakentesting.egnyte.com:443/pubapi/v1/fs-content/Shared/Raken/chao_egnyte_test_project/Raken%20Gallery/01%20BiznessName1/2020/10.%20October/sample.2020.10.15_19.30.40.mp4",
        "method": "POST",
        "headers": {
            "authorization": "Bearer 8sd******************xwq",
            "host": "rakentesting.egnyte.com",
            "content-length": 0,
            "user-agent": "raken-integrations"
        },
        "body": "<0 bytes binary>"
    },
    "2": {
        "statusCode": 403,
        "statusMessage": "Forbidden",
        "headers": {
            "x-mashery-responder": "sjc-gcp-mashery443-slave01",
            "x-accesstoken-qps-allotted": "2",
            "x-accesstoken-qps-current": "3",
            "retry-after": "1",
            "x-error-detail-header": "Account Over Queries Per Second Limit",
            "x-mashery-error-code": "ERR_403_DEVELOPER_OVER_QPS",
            "content-type": "text/xml",
            "date": "Wed, 14 Apr 2021 01:59:14 GMT",
            "content-length": "27",
            "connection": "close",
            "x-content-type-options": "nosniff",
            "x-xss-protection": "1; mode=block",
            "strict-transport-security": "max-age=31536000; includeSubDomains",
            "x-egnyte-request-id": "3426FE37:A019_23F4C2EA:01BB_--_1EC09B|sjc-gcp-glb-l1webui03+http_l1_webui",
            "egnyte-processing-location": "GCP"
        },
        "body": "<27 bytes binary>"
    },
    "service": "raken-integrations",
    "host": "169.254.22.9",
    "timestamp": "2021-04-14T01:59:05.221Z"
}
2021-04-14T01:59:14.715Z	4d82ade8-e019-5427-8ac4-f518aea99fc7	WARN	developer over QPS, retrying
info: 🔵 POST https://rakentesting.egnyte.com:443/pubapi/v1/fs-content/Shared/Raken/chao_egnyte_test_project/Raken%20Gallery/01%20BiznessName1/2020/10.%20October/sample.2020.10.15_19.30.40.mp4 {
    "0": "3e76b74b-ee5e-4601-a17c-a8734e8d22d2",
    "1": {
        "url": "https://rakentesting.egnyte.com:443/pubapi/v1/fs-content/Shared/Raken/chao_egnyte_test_project/Raken%20Gallery/01%20BiznessName1/2020/10.%20October/sample.2020.10.15_19.30.40.mp4",
        "method": "POST",
        "headers": {
            "authorization": "Bearer 8sd******************xwq",
            "host": "rakentesting.egnyte.com",
            "content-length": 0,
            "user-agent": "raken-integrations"
        },
        "body": "<0 bytes binary>"
    },
    "service": "raken-integrations",
    "host": "169.254.22.9",
    "timestamp": "2021-04-14T01:59:05.221Z"
}
info: 🟢 200 POST https://rakentesting.egnyte.com:443/pubapi/v1/fs-content/Shared/Raken/chao_egnyte_test_project/Raken%20Gallery/01%20BiznessName1/2020/10.%20October/sample.2020.10.15_19.30.40.mp4 (231 ms) {
    "0": "3e76b74b-ee5e-4601-a17c-a8734e8d22d2",
    "1": {
        "url": "https://rakentesting.egnyte.com:443/pubapi/v1/fs-content/Shared/Raken/chao_egnyte_test_project/Raken%20Gallery/01%20BiznessName1/2020/10.%20October/sample.2020.10.15_19.30.40.mp4",
        "method": "POST",
        "headers": {
            "authorization": "Bearer 8sd******************xwq",
            "host": "rakentesting.egnyte.com",
            "content-length": 0,
            "user-agent": "raken-integrations"
        },
        "body": "<0 bytes binary>"
    },
    "2": {
        "statusCode": 200,
        "statusMessage": "OK",
        "headers": {
            "x-mashery-responder": "sjc-gcp-mashery443-slave01",
            "x-accesstoken-qps-allotted": "2",
            "x-accesstoken-qps-current": "1",
            "x-accesstoken-quota-allotted": "1000",
            "x-accesstoken-quota-current": "14",
            "x-quota-reset": "Thursday, April 15, 2021 12:00:00 AM GMT",
            "etag": "5ec148b0-7dd8-4e63-bb23-08d2fefb72d4",
            "x-sha512-checksum": "cf83e1357eefb8bdf1542850d66d8007d620e4050b5715dc83f4a921d36ce9ce47d0d13c5d85f2b0ff8318d2877eec2f63b931bd47417a81a538327af927da3e",
            "last-modified": "Wed, 14 Apr 2021 01:59:15 GMT",
            "content-type": "application/json;charset=UTF-8",
            "date": "Wed, 14 Apr 2021 01:59:15 GMT",
            "connection": "close",
            "x-content-type-options": "nosniff",
            "x-xss-protection": "1; mode=block",
            "strict-transport-security": "max-age=31536000; includeSubDomains",
            "x-egnyte-request-id": "3426FE37:FF99_23F4C2EA:01BB_--_1EC9B1|sjc-gcp-glb-l1webui03+http_l1_webui",
            "egnyte-processing-location": "GCP"
        },
        "body": "{\"checksum\":\"cf83e1357eefb8bdf1542850d66d8007d620e4050b5715dc83f4a921d36ce9ce47d0d13c5d85f2b0ff8318d2877eec2f63b931bd47417a81a538327af927da3e\",\"path\":\"/Shared/Raken/chao_egnyte_test_project/Raken Gallery/01 BiznessName1/2020/10. October/sample.2020.10.15_19.30.40.mp4\",\"group_id\":\"d201a8d3-8ae9-4a64-b73c-bbfdf6152d9c\",\"entry_id\":\"5ec148b0-7dd8-4e63-bb23-08d2fefb72d4\"}"
    },
    "service": "raken-integrations",
    "host": "169.254.22.9",
    "timestamp": "2021-04-14T01:59:05.221Z"
}
```

**Comment by uhlryk** [comment]
@chaoyangnz thanks, 
I will create an internal ticket and I will give you ETA.

**Comment by PINTEGNYTE** [comment]
Fix for this issue is provided in version v2.9.0 released in May 7, 2021. In case reaching QPS the storeFile() method will throw error and stop uploading.
